pytorch_lightning/plugins/training_type/training_type_plugin.py

```diff
@@ -41,6 +41,32 @@ def __init__(self) -> None:
         self._results: Optional[Union[_EVALUATE_OUTPUT, _PREDICT_OUTPUT]] = None
         self._call_configure_sharded_model_hook = True
 
+    def pre_dispatch(self) -> None:
+        """Hook to do something before the training/evaluation/prediction starts."""
+
+    def post_dispatch(self) -> None:
+        """Hook to do something after the training/evaluation/prediction finishes."""
+
+    @contextlib.contextmanager
```

Comment:
We use these hooks in `PrecisionPlugin` too so maybe also add them there?

Reply:
It's true, I forgot 🤦 
Then my suggestion was not very clean actually

Reply:
No worries. For now we can explicitly list Plugin and ClusterEnvironment in trainer.plugins type hint, while keeping ClusterEnvironment it's separate ABC